.github/workflows/linux_ci.yml

```diff
@@ -185,11 +251,22 @@ jobs:
     - name: Install tesseract
       run: sudo apt-get install tesseract-ocr libtesseract-dev poppler-utils
 
+    - name: Install Dependencies (on cache miss only)
+      if: steps.cache.outputs.cache-hit != 'true'
+      run: |
+        pip install --upgrade pip
+        pip install .[test]
+        pip install rest_api/
+        pip install ui/
+        pip install torch-scatter -f https://data.pyg.org/whl/torch-1.10.0+cpu.html
+        echo "=== pip freeze ==="
+        pip freeze
+
```

Comment:
Why do we need these blocks in all build-cache descendents? Shouldn't the cache hit always?

Reply:
Yes... the problem is that it "should". The documentation for the cache GitHub action mentions that caches might disappear at any point and recommends to always have a step that runs in case of a cache miss. I also observed several, apparently random cache misses which I could not explain in any way, so for extra safety I added these blocks.

Ugly, I agree. I'll add comments explaining this.